feat: Argument Passing 구현

hyunmuam opened the pull request:
## 🔗 관련 이슈
<!-- 이슈 번호를 연결해주세요 -->
closes #19

## 📌 작업 내용
### 구현 사항
<!-- 구현내용에 대해서 설명해주세요. -->
<!-- 예시 :
- 이슈 템플릿, PR 템플릿 작성
- READEME 파일 최초 작성
 --> 
- **Argument Passing 구현**
  - `file_name`을 파싱하여 프로그램 인자를 사용자 스택에 배치
  - x86-64 SysV ABI 호출 규약에 따라 레지스터 설정
- **리팩토링**
  - `load()` 함수의 책임 분리로 가독성 및 유지보수성 향상
  - 지역 배열을 동적 할당으로 변경하여 커널 스택 오버플로우 방지

### 변경된 파일
<!-- 구현 내용으로 인해 변경된 파일에 대해 설명해주세요 -->
<!-- 예시:
- README.md
  - readme 파일 최초 생성
- issue_template.md
  - 이슈 템플릿 최초 생성
-->
- `userprog/process.c`
  - `parse_arguments()` 함수 추가
    - 명령줄을 공백 기준으로 파싱하여 argc/argv 구성
    - `palloc_get_page()`로 동적 할당하여 커널 스택 부담 해소
  - `setup_argument_stack()` 함수 추가
    - 인자 문자열을 역순으로 사용자 스택에 복사
    - 스택 포인터 8바이트 정렬 (x86-64 호출 규약)
    - argv 포인터 배열 및 NULL terminator 배치
    - rdi(argc), rsi(argv) 레지스터 설정
    - fake return address 추가
  - `load()` 함수 수정
    - 인자 처리 로직을 위 함수들로 위임
    - `filesys_open()` 호출 시 전체 명령줄 대신 `argv[0]` 전달
- `threads/init.c`
  - 코드 컨벤션에 맞춰 포매팅 적용
### 추가 사항
<!-- 트러블 슈팅 등 추가로 설명할 내용이 있으시다면 아래에 작성해주세요. -->

## ✅ 셀프 체크리스트
<!-- 체크리스트를 모두 완료한 후 제출해주세요 -->
<!-- 완료 처리를 표시하려면 [x], 완료되지 않았다면 [ ] -->
- [x] 코드 컨벤션을 준수했나요?
- [x] 불필요한 주석이나 디버깅 코드를 제거했나요?
- [x] 관련 테스트가 모두 통과하나요?
- [x] 커밋 메시지가 컨벤션에 맞나요?

hyunmuam committed:
style: init.c, process.c 포매팅 규칙 적용

hyunmuam committed:
feat: Argument Passing 구현
- strtok_r로 공백 기준 file_name 파싱하여 argc/argv 구성
- 인자 문자열을 역순으로 사용자 스택에 복사 및 주소 업데이트
- x86-64 호출 규약에 따라 스택 8바이트 정렬
- argv 포인터 배열과 NULL terminator를 스택에 배치
- rdi(argc), rsi(argv) 레지스터 설정 및 fake return address 추가

hyunmuam committed:
refactor: load()에서 인자 처리 로직 분리 및 동적 할당

배경:
- 기존 load() 함수가 ELF 로딩, 인자 파싱, 스택 설정을 모두 담당하여 가독성이 떨어지고 책임이 과도하게 집중됨
- 지역 변수(fn_copy[128], argv[64])가 커널 스택을 차지하여 스택 오버플로우 위험 존재

변경 사항:
- parse_arguments(): 명령줄 파싱 및 argv/argc 구성 담당
- setup_argument_stack(): x86-64 SysV ABI에 맞게 사용자 스택 적재
- fn_copy, argv를 palloc_get_page()로 동적 할당하여 커널 스택 부담 해소

hdh853 commented on userprog/process.c:
```diff
@@ -441,48 +411,91 @@ static bool load(const char* file_name, struct intr_frame* if_)
-        memcpy((void*)if_->rsp, &argv[i], sizeof(uint64_t));
+/* 명령줄을 파싱하여 argv/argc 구성
+   성공 시 true 반환, 실패 시 false 반환
+   주의: fn_copy는 호출자가 해제해야 함 */
```
@tjvm0877 
fn_copy를 호출자가 해제해야한다고 하셨는데, 따로 해제하는 로직은 보이지 않아 어디서 해제하고 계신건지 궁금합니다!

tjvm0877 replied:
@hdh853 
원래 `fn_copy`는 `load`에서 넘겨주는 변수였는데, 명령줄 파싱 외에는 사용하지 않아 `parse_arguments`로 넣었는데 그 과정에서 `free` 호출이 누락된 것 같네요.

hyunmuam replied:
원래 `fn_copy`는 `load()`에서 관리하던 변수였는데, 
리팩토링하면서 `parse_arguments()`로 옮기는 과정에서 
수명 관리를 잘못파악한 것 같습니다.

다시 확인해보니 `argv[]`의 모든 포인터가 `fn_copy` 내부를 가리키므로,
`fn_copy`를 먼저 해제하면 dangling pointer가 됩니다.

다음과 같이 수정하려합니다.
- `fn_copy`는 `load()` 스코프에서 관리
- `parse_arguments()`에서는 파싱만 수행
- 스택 설정 후 `fn_copy` 해제

hdh853 commented on userprog/process.c:
```diff
@@ -441,48 +411,91 @@ static bool load(const char* file_name, struct intr_frame* if_)
+    char* fn_copy = palloc_get_page(PAL_ZERO);
+    if (fn_copy == NULL) {
+        palloc_free_page(fn_copy);
+        return false;
```
fn_copy가 널일때, palloc_free_page(NULL)이 되어 버려서 커널패닉이 발생할 것 같습니다. 이럴경우 할당자체가 실패한 경우이니 return false만 해줘도 괜찮지 않을까 싶습니다!

hyunmuam replied:
넵 지워줘야합니다.

hdh853 commented on userprog/process.c:
```diff
@@ -441,48 +411,91 @@ static bool load(const char* file_name, struct intr_frame* if_)
-    if_->R.rdi = argc;            // 첫 번째 인자: argc
-    if_->R.rsi = if_->rsp + 8;    // 두 번째 인자: argv
+    /* argv용 별도의 4KB 페이지 할당 */
+    char** argv = palloc_get_page(PAL_ZERO);
+    if (argv == NULL) {
+        palloc_free_page(argv);
+        return false;
```
여기도 마찬가지로 palloc_free_page(NULL)이 되어버릴 수 있을것 같습니다

hyunmuam replied:
여기는 ` palloc_free_page(fn_copy);`로 바꿔주면 될 것 같습니다.

hdh853 reviewed (commented):
argument passing 로직을 외부 함수로 빼놓으니 load함수도 가벼워지고 가독성도 좋아지는것 같습니다. 고생하셨습니다!

hyunmuam committed:
fix: fn_copy 메모리 수명 관리 오류 수정

문제:
- parse_arguments()에서 할당한 fn_copy를 조기 해제 시 argv[] 포인터들이 dangling pointer가 되는 문제

원인:
- 리팩토링 과정에서 fn_copy 수명 관리 책임이 불명확해짐
- argv[i]는 fn_copy 내부 토큰을 가리키므로 fn_copy보다 먼저 해제되면 안 됨

수정:
- fn_copy를 load() 스코프에서 관리하도록 변경
- parse_arguments()는 파싱만 수행
- setup_argument_stack() 완료 후 fn_copy 해제

hdh853 reviewed (approved):
고생하셨습니다~
